src/main/scala/mesosphere/marathon/api/LeaderProxyFilter.scala

```diff
@@ -1,137 +1,263 @@
-    // TODO handle https
+  private[this] def buildUrl(leaderData: String, request: HttpServletRequest): URL = {
     if (request.getQueryString != null) {
-      new URL("http://%s%s?%s".format(leaderData, request.getRequestURI, request.getQueryString))
+      new URL(s"$scheme://$leaderData${request.getRequestURI}?${request.getQueryString}")
     }
     else {
-      new URL("http://%s%s".format(leaderData, request.getRequestURI))
+      new URL(s"$scheme://$leaderData${request.getRequestURI}")
     }
   }
 
-  def doFilter(rawRequest: ServletRequest,
-               rawResponse: ServletResponse,
-               chain: FilterChain) {
+  @tailrec
+  final def doFilter(rawRequest: ServletRequest,
+                     rawResponse: ServletResponse,
+                     chain: FilterChain) {
 
-    if (rawRequest.isInstanceOf[HttpServletRequest]) {
-      val request = rawRequest.asInstanceOf[HttpServletRequest]
-      val response = rawResponse.asInstanceOf[HttpServletResponse]
+    def waitForConsistentLeadership(response: HttpServletResponse): Boolean = {
+      var retries = 10
 
-      // Proxying occurs if we aren't in the leadership position and we know about the other leader (to proxy to).
-      if (schedulerService.isLeader || schedulerService.getLeader.isEmpty) {
-        chain.doFilter(request, response)
-      }
-      else {
-        try {
-          val leaderData = schedulerService.getLeader
+      do {
+        val weAreLeader = leaderInfo.elected
+        val currentLeaderData = leaderInfo.currentLeaderHostPort()
 
-          log.info(s"Proxying request to leader at ${leaderData.get}")
+        if (weAreLeader || currentLeaderData.exists(_ != myHostPort)) {
+          log.info("Leadership info is consistent again!")
+          return true
+        }
 
-          val method = request.getMethod
+        if (retries >= 0) {
+          log.info(s"Waiting for consistent leadership state. Are we leader?: $weAreLeader, leader: $currentLeaderData")
+          sleep()
+        }
+        else {
+          log.error(
+            s"inconsistent leadership state, refusing request for ourselves at $myHostPort. " +
+              s"Are we leader?: $weAreLeader, leader: $currentLeaderData")
+        }
 
-          val proxy =
-            buildUrl(leaderData.get, request)
-              .openConnection().asInstanceOf[HttpURLConnection]
+        retries -= 1
+      } while (retries >= 0)
 
-          // getHeaderNames() and getHeaders() are known to return null, see:
-          //http://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeaders(java.lang.String)
-          val names = Option(request.getHeaderNames).map(_.asScala).getOrElse(Nil)
+      false
+    }
 
-          for (name <- names) {
-            val values = request.getHeaders(name)
-            if (values != null) {
-              proxy.setRequestProperty(name, values.asScala.mkString(","))
-            }
+    (rawRequest, rawResponse) match {
```

Comment:
There is a lot of nested if/else statements. Can you define some helper methods and make the main control flow more obvious?